RavenDB-23346 Fixing LazyStringParser.TryParseDateTime to take into account that trailing zeros can be dropped when using .FFFFFFFK / .FFFK format. Adding support for r format specifier

arekpalinski opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-23346/LazyStringParser.TryParseDateTime-doesnt-handle-trailing-zeros-properly-and-r-format

https://issues.hibernatingrhinos.com/issue/RavenDB-21634/SlowTests.Server.Documents.Expiration.ExpirationTests.CanAddEntityWithExpiryThenReadItBeforeItExpiresButWillNotBeAbleToReadItAft

### Additional description

The problem is that when using `F` as fraction specifier in date format then trailing zeros are dropped (in contrast to `f`).

https://learn.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings#uppercase-seconds-f-fraction-specifier

We use `F` in our supported formats for reads:
https://github.com/ravendb/ravendb/blob/76149a7367434ff60ff9116aaf7d0dbf0f821a56/src/Sparrow/DefaultFormat.cs#L19-L26

This PR makes LazyStringParser aware of that. Also it adds support for "r" format via regular `DateTimeOffset.TryParse`.

It adds support for parsing dates with `.FFFFFFFK` but also for `.FFFFFFF` while this format is not defined in `DateTimeFormatsToRead`. Although we did add support for `.fff` (note no `K` at the end) in https://github.com/ravendb/ravendb/pull/13287 / https://github.com/ravendb/ravendb/pull/13313 so I believe it would be hard to explain why parsing fractional date with 3 digits works while it's not the case for 2 or 4 digits (for none UTC time).

@redknightlois here is how the benchmark looks after changes:

```
| Method                        | Mean     | Error    | StdDev   | StdErr   | Min      | Q1       | Median   | Q3       | Max      | Op/s         |
|------------------------------ |---------:|---------:|---------:|---------:|---------:|---------:|---------:|---------:|---------:|-------------:|
| TryParseDateTimeBenchmark     | 14.43 ns | 0.270 ns | 0.252 ns | 0.065 ns | 14.14 ns | 14.19 ns | 14.52 ns | 14.69 ns | 14.79 ns | 69,277,521.1 |
| TryParseDateTimeBenchmark_Old | 13.32 ns | 0.160 ns | 0.150 ns | 0.039 ns | 13.11 ns | 13.15 ns | 13.37 ns | 13.42 ns | 13.57 ns | 75,089,820.0 |
```

TODO: 

- this change might affect indexed terms - in https://github.com/ravendb/ravendb/pull/13313 we introduced `IndexVersion.ProperlyParseThreeDigitsMillisecondsDates` - we might need the same here

### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [x] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

TODO

- [ ] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [ ] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [x] Yes. Please list the affected features/subsystems and provide appropriate explanation - Indexing, Expiration, Refresh, Archival
- [ ] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

arekpalinski committed:
RavenDB-23346 Fixing LazyStringParser.TryParseDateTime to take into account that trailing zeros can be dropped when using .FFFFFFFK / .FFFK format. Adding support for r format specifier

redknightlois commented on src/Sparrow/Json/LazyStringParser.cs:
```diff
@@ -683,7 +685,17 @@ public static Result TryParseDateTime(byte* buffer, int len, out DateTime dt, ou
-                goto Failed;
+            {
+                if (len != 29 || buffer[3] != ',' || buffer[19] != ':' || buffer[22] != ':')
+                    goto Failed;
```
All these checks can be done with SIMD instructions now. I wouldn't worry much about the difference and just create an issue to tackle this eventually.  Unless performance test in real world scenarios show important performance regressions.

arekpalinski replied:
@redknightlois can you please create it? You know best how to formulate / describe it. 

Based on the above benchmark results (14.43 ns vs 13.32 ns per operation) , do you think this PR is acceptable for 5.4? 
The SIMD optimization could be targeted to 6.x / 7.x if not really needed in 5.4.

redknightlois replied:
It's fine for 6.x imho.  https://issues.hibernatingrhinos.com/issue/RavenDB-23390/Analyze-TryParseDateTime-performance-with-SIMD-instructions

ppekrol commented on test/FastTests/Utils/TimeParsing.cs:
```diff
@@ -151,7 +154,130 @@ public void TimeOnly(string date, int hh, int mm, int ss, int ms)
+            var noOfDaysInMonth = DateTime.DaysInMonth(year, month);
+            var day = random.Next(1, noOfDaysInMonth);
+
+            return new DateTime(year, month, day);
```
shouldn't we put milliseconds here? if not then it is set to 0:00.00 then?

arekpalinski replied:
Yes. Fixed

ppekrol commented on test/FastTests/Utils/TimeParsing.cs:
```diff
@@ -151,7 +154,130 @@ public void TimeOnly(string date, int hh, int mm, int ss, int ms)
+                {
+                    var parseResult = LazyStringParser.TryParseDateTime(buffer, bytes.Length, out var dateTime, out var dateTimeOffset, properlyParseThreeDigitsMilliseconds: true);
+                    
+                    Assert.NotEqual(LazyStringParser.Result.Failed, parseResult);
```
I would add the 'tested' value to the message here

arekpalinski replied:
Changed to Assert.True and added a message

ppekrol commented on test/FastTests/Utils/TimeParsing.cs:
```diff
@@ -151,7 +154,130 @@ public void TimeOnly(string date, int hh, int mm, int ss, int ms)
+
+        [RavenTheory(RavenTestCategory.Core)]
+        [InlineDataWithRandomSeed]
+        public void CanParseValidRandomDate(int seed)
```
can we use that seed value as ticks in the `new DateTime(ticks, kind)` ctor? Thoughts?

arekpalinski replied:
Not every random seed (passed as ticks) is a valid date .

arekpalinski commented:
test this in debug please

arekpalinski committed:
RavenDB-23346 Applying the fix to LazyStringParser.TryParseDateTime but for the overload accepting char*. Added ability to switch to the old implementation for backward compatiblity purposes. Tests added.

arekpalinski committed:
RavenDB-23346 Removed option to fallback to old code. We're not gonna use it anyway since.

arekpalinski commented:
Regarding backward compatibility concerns about indexing / querying, I already mentioned above:

> this change might affect indexed terms - in https://github.com/ravendb/ravendb/pull/13313 we introduced IndexVersion.ProperlyParseThreeDigitsMillisecondsDates - we might need the same here

- in https://github.com/ravendb/ravendb/pull/13313 we have applied the index version check _only_ to the querying part (where we check if a given parameter is a string). This is not a complete handling since new index entries will get new formatting hence this isn't a valid check.
- currently if we have a date like `2021-12-14T11:22:28.3224Z` then on the server side we'll not recognize it as date, hence we won't have `_Ticks` field available. Although it's likely that other values will be detected as `DateTime` so queries will attempt to use `_Ticks` field anyway
- queries doing range checks on fields with `DateTime` values will filter out entries where `_Ticks` fields is missing
- RavenDB clients send `DateTime` values as query parameters with `.fffffffK` formatting so equality checks will not return valid results, even comparison by string wouldn't work

So the conclusion is that fixing this issue has more benefits for indexing and querying than potential backward compatibility problems. Currently documents having date time values like `2021-12-14T11:22:28.3224Z` aren't properly handled and index entries contain mixed terms (string and _Ticks vs only string), hence aren't properly returned in most queries. After this change the new and updated entries will be handled correctly.
